Parser/pegen/pegen.c

```diff
@@ -0,0 +1,1834 @@
+        if (m->type == type) {
+            // Update existing node.
+            m->node = node;
+            m->mark = p->mark;
+            return 0;
+        }
+    }
+    // Insert new node.
+    return _PyPegen_insert_memo(p, mark, type, node);
+}
+
+// Return dummy NAME.
+void *
+_PyPegen_dummy_name(Parser *p, ...)
+{
+    static void *cache = NULL;
+
+    if (cache != NULL) {
+        return cache;
+    }
+
+    PyObject *id = _create_dummy_identifier(p);
+    if (!id) {
+        return NULL;
+    }
+    cache = Name(id, Load, 1, 0, 1, 0, p->arena);
+    return cache;
+}
+
+static int
+_get_keyword_or_name_type(Parser *p, const char *name, int name_len)
+{
+    if (name_len >= p->n_keyword_lists || p->keywords[name_len] == NULL) {
+        return NAME;
+    }
+    for (KeywordToken *k = p->keywords[name_len]; k->type != -1; k++) {
+        if (strncmp(k->str, name, name_len) == 0) {
+            return k->type;
+        }
+    }
+    return NAME;
+}
+
+int
+_PyPegen_fill_token(Parser *p)
+{
+    const char *start, *end;
+    int type = PyTokenizer_Get(p->tok, &start, &end);
+    if (type == ERRORTOKEN) {
+        return tokenizer_error(p);
+    }
+    if (type == ENDMARKER && p->start_rule == Py_single_input && p->parsing_started) {
+        type = NEWLINE; /* Add an extra newline */
+        p->parsing_started = 0;
+    }
+    else {
+        p->parsing_started = 1;
+    }
+
+    if (p->fill == p->size) {
+        int newsize = p->size * 2;
+        p->tokens = PyMem_Realloc(p->tokens, newsize * sizeof(Token *));
+        if (p->tokens == NULL) {
+            PyErr_Format(PyExc_MemoryError, "Realloc tokens failed");
+            return -1;
+        }
+        for (int i = p->size; i < newsize; i++) {
+            p->tokens[i] = PyMem_Malloc(sizeof(Token));
+            memset(p->tokens[i], '\0', sizeof(Token));
+        }
+        p->size = newsize;
+    }
+
+    Token *t = p->tokens[p->fill];
+    t->type = (type == NAME) ? _get_keyword_or_name_type(p, start, (int)(end - start)) : type;
+    t->bytes = PyBytes_FromStringAndSize(start, end - start);
+    if (t->bytes == NULL) {
+        return -1;
+    }
+    PyArena_AddPyObject(p->arena, t->bytes);
+
+    int lineno = type == STRING ? p->tok->first_lineno : p->tok->lineno;
+    const char *line_start = type == STRING ? p->tok->multi_line_start : p->tok->line_start;
+    int end_lineno = p->tok->lineno;
+    int col_offset = -1, end_col_offset = -1;
+    if (start != NULL && start >= line_start) {
+        col_offset = start - line_start;
+    }
+    if (end != NULL && end >= p->tok->line_start) {
+        end_col_offset = end - p->tok->line_start;
+    }
+
+    t->lineno = lineno;
+    t->col_offset = col_offset;
+    t->end_lineno = end_lineno;
+    t->end_col_offset = end_col_offset;
+
+    // if (p->fill % 100 == 0) fprintf(stderr, "Filled at %d: %s \"%s\"\n", p->fill,
+    // token_name(type), PyBytes_AsString(t->bytes));
```

Comment:
Maybe remove this debug code, or add an #ifdef? Same remark for commented fprintf below.

PyConfig has a parser_debug member. I don't know who use it :-)

Reply:
Just removed them.